sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/planning/SelectedFieldSuite.scala

```diff
@@ -0,0 +1,440 @@
+  }
+
+  test("col4['foo'].field1.subfield1 as foo") {
+    val expected =
+      StructField("col4", MapType(StringType, StructType(
+        StructField("field1", StructType(
+          StructField("subfield1", IntegerType, nullable = false) :: Nil)) :: Nil), valueContainsNull = false), nullable = false)
+    testSelect("col4['foo'].field1.subfield1 as foo", expected)
+  }
+
+  test("col4['foo'].field2['bar'] as foo") {
+    val expected =
+      StructField("col4", MapType(StringType, StructType(
+        StructField("field2", MapType(StringType, IntegerType, valueContainsNull = false)) :: Nil), valueContainsNull = false), nullable = false)
+    testSelect("col4['foo'].field2['bar'] as foo", expected)
+  }
+
+  test("col5[0]['foo'].field1.subfield1 as foo") {
+    val expected =
+      StructField("col5", ArrayType(MapType(StringType, StructType(
+        StructField("field1", StructType(
+          StructField("subfield1", IntegerType) :: Nil)) :: Nil), valueContainsNull = false)))
+    testSelect("col5[0]['foo'].field1.subfield1 as foo", expected)
+  }
+
+  test("col6['foo'][0].field1.subfield1 as foo") {
+    val expected =
+      StructField("col6", MapType(StringType, ArrayType(StructType(
+        StructField("field1", StructType(
+          StructField("subfield1", IntegerType) :: Nil)) :: Nil), containsNull = false)))
+    testSelect("col6['foo'][0].field1.subfield1 as foo", expected)
+  }
+
+  test("col2.field5.subfield1.subsubfield1") {
+    val expected =
+      StructField("col2", StructType(
+        StructField("field5", ArrayType(StructType(
+          StructField("subfield1", StructType(
+            StructField("subsubfield1", IntegerType) :: Nil), nullable = false) :: Nil)), nullable = false) :: Nil))
+    testSelect("col2.field5.subfield1.subsubfield1", expected)
+  }
+
+  test("col2.field5.subfield2.subsubfield1.subsubsubfield1") {
+    val expected =
+      StructField("col2", StructType(
+        StructField("field5", ArrayType(StructType(
+          StructField("subfield2", StructType(
+            StructField("subsubfield1", StructType(
+              StructField("subsubsubfield1", StringType) :: Nil)) :: Nil)) :: Nil)), nullable = false) :: Nil))
+    testSelect("col2.field5.subfield2.subsubfield1.subsubsubfield1", expected)
+  }
+
+  test("col2.field4['foo'].subfield1 as foo") {
+    val expected =
+      StructField("col2", StructType(
+        StructField("field4", MapType(StringType, StructType(
+          StructField("subfield1", IntegerType) :: Nil), valueContainsNull = false)) :: Nil))
+    testSelect("col2.field4['foo'].subfield1 as foo", expected)
+  }
+
+  test("col2.field4['foo'].subfield2 as foo, col2.field4['foo'].subfield2[0] as foo") {
+    val expected =
+      StructField("col2", StructType(
+        StructField("field4", MapType(StringType, StructType(
+          StructField("subfield2", ArrayType(IntegerType, containsNull = false)) :: Nil), valueContainsNull = false)) :: Nil))
```

Comment:
I feel those expected schema are hard to read. Can we have a helper function that returns part of the `schema` based on what fields requested?

Reply:
Then we don't need to set style off for the whole test file.

Reply:
It's something I'll look into.